src/networking/payloads/introductionrequestpayload.rs

```diff
@@ -32,125 +34,70 @@ struct IntroductionRequestPayload {
-    let flags = packetiter.next_bits().unwrap();
-
-    let identifier = packetiter.next_u16().unwrap();
-
-    let extra_bytes = packetiter.next_raw_remaining().unwrap();
+/// makes the IntroductionRequestPayload serializable.
+/// This is less than trivial as there is no 1:1 mapping between the serialized data and the payload struct.
+/// Some struct fields are combined into one byte to form the serialized data.
+impl Serialize for IntroductionRequestPayload {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer{
+        let conntype = self.connection_type.encode();
+
+        let mut state = serializer.serialize_struct("IntroductionRequestPayload", 6)?;
+        state.serialize_field("destination_address", &self.destination_address)?;
+        state.serialize_field("source_lan_address", &self.source_lan_address)?;
+        state.serialize_field("source_wan_address", &self.source_wan_address)?;
+        state.serialize_field("advice", &Bits::from_bools((conntype.0, conntype.1, false, false, false, false, false, self.advice)))?;
+        state.serialize_field("identifier", &self.identifier)?;
+        state.serialize_field("extra_bytes", &self.extra_bytes)?;
+
+        state.end()
+    }
+}
 
-    IntroductionRequestPayload {
-      destination_address: Address {
-        address: Ipv4Addr::new(
-          destination_ipaddress[0],
-          destination_ipaddress[1],
-          destination_ipaddress[2],
-          destination_ipaddress[3],
-        ),
-        port: destination_port,
-      },
-      source_lan_address: Address {
-        address: Ipv4Addr::new(
-          source_lan_address[0],
-          source_lan_address[1],
-          source_lan_address[2],
-          source_lan_address[3],
-        ),
-        port: source_lan_port,
-      },
-      source_wan_address: Address {
-        address: Ipv4Addr::new(
-          source_wan_address[0],
-          source_wan_address[1],
-          source_wan_address[2],
-          source_wan_address[3],
-        ),
-        port: source_wan_port,
-      },
-      advice: flags.bit7,
-      connection_type: ConnectionType::decode((flags.bit0, flags.bit1)),
-      identifier,
-      extra_bytes,
+#[derive(Debug, PartialEq, serde::Deserialize)]
+/// this is the actual pattern of an introductionRequestPayload.
+/// Used for deserializing. This is again needed because there is no 1:1 mapping between the
+/// serialized data and the payload struct. This is the intermediate representation.
+struct IntroductionRequestPayloadPattern(Address,Address,Address,Bits,u16);
+
+impl<'de> Deserialize<'de> for IntroductionRequestPayload{
+  /// deserializes an IntroductionRequestPayload
+  fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+  where D: Deserializer<'de>,{
+    // first deserialize it to a temporary struct which literally represents the packer
+    let temppayload = IntroductionRequestPayloadPattern::deserialize(deserializer);
```

Comment:
`temppayload` looks like a typo, but it isn't. Please, consider using CamelCase or snake_case for two-words identifiers. And it is better to put adjective second because nouns are more generic, and our brains are more used to see generic words first. Think animal taxonomy: `canis_lupus`.